appcues/src/main/java/com/appcues/Storage.kt

```diff
@@ -21,31 +23,79 @@ internal class Storage(
-        set(value) = sharedPreferences.edit().putString(Constants.DeviceId.rawVal, value).apply()
+        get() = _deviceId
+        set(value) = updateDeviceId(value)
+
+    private var _deviceId: String
+
+    private fun updateDeviceId(deviceId: String) {
+        _deviceId = deviceId
+        sharedPreferences.edit().putString(Constants.DeviceId.rawVal, _deviceId).apply()
+    }
 
     var userId: String
-        get() = sharedPreferences.getString(Constants.UserId.rawVal, null) ?: ""
-        set(value) = sharedPreferences.edit().putString(Constants.UserId.rawVal, value).apply()
+        get() = _userId
+        set(value) = updateUserId(value)
+
+    private var _userId: String
+
+    private fun updateUserId(userId: String) {
+        _userId = userId
+        sharedPreferences.edit().putString(Constants.UserId.rawVal, _userId).apply()
+    }
 
     var groupId: String?
-        get() = sharedPreferences.getString(Constants.GroupId.rawVal, null)
-        set(value) = sharedPreferences.edit().putString(Constants.GroupId.rawVal, value).apply()
+        get() = _groupId
+        set(value) = updateGroupId(value)
+
+    private var _groupId: String? = null
+
+    private fun updateGroupId(groupId: String?) {
+        _groupId = groupId
+        sharedPreferences.edit().putString(Constants.GroupId.rawVal, _groupId).apply()
+    }
 
     var isAnonymous: Boolean
-        get() = sharedPreferences.getBoolean(Constants.IsAnonymous.rawVal, true)
-        set(value) = sharedPreferences.edit().putBoolean(Constants.IsAnonymous.rawVal, value).apply()
+        get() = _isAnonymous
+        set(value) = updateIsAnonymous(value)
+
+    private var _isAnonymous: Boolean = true
+
+    private fun updateIsAnonymous(isAnonymous: Boolean) {
+        _isAnonymous = isAnonymous
+        sharedPreferences.edit().putBoolean(Constants.IsAnonymous.rawVal, _isAnonymous).apply()
+    }
 
     var lastContentShownAt: Date?
-        get() = sharedPreferences.getLong(Constants.LastContentShownAt.rawVal, 0).run { if (this > 0) Date(this) else null }
-        set(value) = when (value) {
-            null -> sharedPreferences.edit().remove(Constants.LastContentShownAt.rawVal).apply()
-            else -> sharedPreferences.edit().putLong(Constants.LastContentShownAt.rawVal, value.time).apply()
+        get() = _lastContentShownAt
+        set(value) = updateLastContentShownAt(value)
+
+    private var _lastContentShownAt: Date? = null
+
+    private fun updateLastContentShownAt(lastContentShownAt: Date?) {
+        _lastContentShownAt = lastContentShownAt
+        _lastContentShownAt?.let {
+            sharedPreferences.edit().putLong(Constants.LastContentShownAt.rawVal, it.time).apply()
+        } ?: run {
+            sharedPreferences.edit().remove(Constants.LastContentShownAt.rawVal).apply()
         }
+    }
 
     init {
-        if (deviceId.isEmpty()) {
-            deviceId = UUID.randomUUID().toString()
+        val oldPolicy = StrictMode.allowThreadDiskReads()
```

Comment:
if we're going to go the route of allowing for now, what do you think of the approach here to simplify the code a bit and avoid having all the backing fields for properties? https://stackoverflow.com/a/49843629

idea would be to have a helper like 
```
fun <T> allowReads(block: () -> T): T {
    val oldPolicy = StrictMode.allowThreadDiskReads()
    try {
        return block()
    } finally {
        StrictMode.setThreadPolicy(oldPolicy)
    }
}
```

then the old line like 
```
get() = sharedPreferences.getString(Constants.UserId.rawVal, null) ?: ""
```

just becomes
```
get() = allowReads { sharedPreferences.getString(Constants.UserId.rawVal, null) ?: "" }
```

Reply:
will do